core/checksums/src/main/java/software/amazon/awssdk/checksums/internal/DigestAlgorithm.java

```diff
@@ -27,41 +31,92 @@ public enum DigestAlgorithm {
     SHA256("SHA-256")
     ;
 
+    private static final int MAX_CACHED_DIGESTS = 10_000;
     private final String algorithmName;
-    private final DigestThreadLocal digestReference;
+    private final Deque<MessageDigest> digestCache = new ConcurrentLinkedDeque<>(); // Used as LIFO for cache-friendliness
 
     DigestAlgorithm(String algorithmName) {
         this.algorithmName = algorithmName;
-        digestReference = new DigestThreadLocal(algorithmName);
     }
 
     public String getAlgorithmName() {
         return algorithmName;
     }
 
     /**
-     * Returns the thread local reference for the {@link MessageDigest} algorithm
+     * Returns a {@link CloseableMessageDigest} to use for this algorithm.
      */
-    public MessageDigest getDigest() {
-        MessageDigest digest = digestReference.get();
-        digest.reset();
-        return digest;
+    public CloseableMessageDigest getDigest() {
+        MessageDigest digest = digestCache.pollFirst();
+        if (digest != null) {
+            digest.reset();
+            return new CloseableMessageDigest(digest);
+        }
+        return new CloseableMessageDigest(newDigest());
+    }
+
+    private MessageDigest newDigest() {
+        try {
+            return MessageDigest.getInstance(algorithmName);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Unable to fetch message digest instance for Algorithm "
+                                       + algorithmName + ": " + e.getMessage(), e);
+        }
     }
 
-    private static class DigestThreadLocal extends ThreadLocal<MessageDigest> {
-        private final String algorithmName;
+    public final class CloseableMessageDigest implements SdkAutoCloseable, Cloneable {
+        private final Supplier<MessageDigest> closedDigest = () -> {
+            throw new IllegalStateException("This message digest is closed.");
+        };
+        private Supplier<MessageDigest> digest;
+        private byte[] messageDigest;
+
+        private CloseableMessageDigest(MessageDigest digest) {
+            this.digest = () -> digest;
+        }
+
+        /**
+         * Retrieve the message digest instance.
+         */
+        public MessageDigest messageDigest() {
+            return digest.get();
+        }
 
-        DigestThreadLocal(String algorithmName) {
-            this.algorithmName = algorithmName;
+        /**
+         * Retrieve the message digest bytes. This will close the message digest when invoked. This is because the underlying
```

Comment:
Question: where do we close messageDigest in this method?

Reply:
Good catch, that needs a test added.